7 Спринт 1/5

Mur-Muring opened the pull request:
Добавлена реализация технического задания 7 спринта

Mur-Muring committed:
7 Спринт 1/5

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+    private final File file;
+    private static final String title = "id,type,name,status,description,epic";
```
static final константы уже именуются в стиле `UPPER_SNAKE_CASE`

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+
+    public FileBackedTaskManager(File file) {
+        this.file = file;
+        save();
```
При создании менеджера save() вызывать нет необходимость, он зашит в логику нужных методов, которые влияют на содержимое файла и сохранение происходит при вызове этих методов

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+
+    @Override
+    public Task addTask(Task task) {
+        Task task1 = super.addTask(task);
+        save();
+        return task1;
+    }
+
+    @Override
+    public Epic addEpic(Epic epic) {
+        Epic epic1 = super.addEpic(epic);
+        save();
+        return epic1;
+    }
+
+    @Override
+    public Subtask addSubtask(Subtask subtask) {
+        Subtask subtask1 = super.addSubtask(subtask);
+        save();
+        return subtask1;
```
Лучше не использовать в названиях переменных индексы и разные префиксы, по типу task1, task2 и тд.

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        save();
+    }
+
+    @Override
+    public ArrayList<Task> getAllTasks() {
+        save();
+        return super.getAllTasks();
+    }
+
+    @Override
+    public ArrayList<Epic> getAllEpics() {
+        save();
+        return super.getAllEpics();
+    }
+
+    @Override
+    public ArrayList<Subtask> getAllSubtasks() {
+        save();
+        return super.getAllSubtasks();
+    }
```
Методы GetAll...() не должны на содержимое файла, их не нужно переопределять с логикой сохранения в файл. Они просто возвращают список задач

Ivanov-s-tmn commented on src/exception/ManagerSaveException.java:
Собственные исключения стоит делать непроверяемыми (наследниками от `RuntimeException`), такие исключения не обязательно обрабатывать и компилятор за этим не следит

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    private static String toString(Task task) {
+        TypeOfTask typeOfTask = toEnum(task);
+        switch (typeOfTask) {
+            case TASK, EPIC -> {
+                return String.format("%d,%s,%s,%s,%s,", task.getIdTask(), typeOfTask, task.getName(), task.getStatus(),
+                        task.getDescription());
+            }
+            case SUBTASK -> {
+                Subtask subTask = (Subtask) task;
+                return String.format("%d,%s,%s,%s,%s,%d", subTask.getIdTask(), typeOfTask, subTask.getName(),
+                        subTask.getStatus(), subTask.getDescription(), subTask.getIdEpic());
+            }
+        }
+
+        throw new IllegalStateException("Неверный тип задач: " + typeOfTask);
+    }
```
Методы преобразования из строки/в строку стоит вынести в отдельный утилитный класс, чтобы разгрузить менеджер. Методы сделать статическими и вызывать через обращение к классу

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        throw new IllegalStateException("Неверный тип задач: " + typeOfTask);
+    }
+
+    private static TypeOfTask toEnum(Task task) {
+        if (task instanceof Epic) {
+            return TypeOfTask.EPIC;
+        } else if (task instanceof Subtask) {
+            return TypeOfTask.SUBTASK;
+        }
+        return TypeOfTask.TASK;
+    }
```
instanceof лучше не использовать, в java и ООП программах он практически не используется, так как нарушает принципы ООП. Достаточно добавить геттеры в классы задач, которые будут возвращать нужный тип в зависимости от класса, в котором он определен

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    private static Status statusFromString(String string) {
+        return switch (string) {
+            case "NEW" -> Status.NEW;
+            case "IN_PROGRESS" -> Status.IN_PROGRESS;
+            case "DONE" -> Status.DONE;
+            default -> throw new IllegalStateException("Неверное значение: " + string);
+        };
```
Можно легко преобразовать строковое значение енама в константу, `Status status = Status.valueOf(strings[3])`, для этого не нужен дополнительный метод. Есть есть желание сделать проверку на возможные статусы, то это уже делается с помощью метода внутри самого енама

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+                TypeOfTask typeOfTask = toEnum(task);
+
+                switch (typeOfTask) {
+                    case TASK -> fileBackedTaskManager.addTask(task);
+                    case EPIC -> fileBackedTaskManager.addEpic((Epic) task);
+                    case SUBTASK -> fileBackedTaskManager.addSubtask((Subtask) task);
```
Вызов методов добавления при восстановлении файла не подойдет, так как эти методы вызывают сохранение в файл и тем самым изменяют файл

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+                bufferedWriter.write(String.format("\n%s", newString));
+            }
+        } catch (IOException e) {
+            throw new ManagerSaveException(e);
```
Для исключения лучше писать сообщение по типу "Ошибка при записи в файл" или что-то подобное)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,258 @@
+        return super.getAllSubtasks();
+    }
+
+    public void save() {
```
Метод save() стоит закрыть приватным модификатором доступа, он используется в логике других методов менеджера и не нужен нам снаружи

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
Для файлового менеджера достаточно покрыть тестом  восстановления состояния из файла, тест-кейс примерно такой:
1. Создать файловый менеджер
2. Добавить несколько задач
3. Создать новый файловый менеджер из того же файла через loadFromFile()
4. Проверить, что новый файловый менеджер корректно восстановил состояние из файла (списки задач/подзадач/эпиков обоих менеджеров идентичны)

Mur-Muring committed:
7 Спринт 2/5

Изменить метод восстановления из файлв. Добавить утилитарный класс. Добавить тест на восстановления из файла.

Mur-Muring committed:
7 Спринт 2/5

Изменить метод восстановления из файлв. Добавить утилитарный класс. Добавить тест на восстановления из файла.

Mur-Muring committed:
7 Спринт 2/5

Изменить метод восстановления из файлв. Добавить утилитарный класс. Добавить тест на восстановления из файла.

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -11,7 +11,7 @@
-    private Integer idCounter = 0;
+    protected Integer idCounter = 0;
 
     HistoryManager historyManager = Managers.getDefaultHistory();
```
Не забывай про инкапсуляцию и закрывай поля модификаторами доступа

Ivanov-s-tmn commented on src/manager/Utils.java:
```diff
@@ -0,0 +1,62 @@
+
+public class Utils {
+    static String toString(Task task) {
+        TypeOfTask typeOfTask = toEnum(task);
+        switch (typeOfTask) {
```
Смотри, ты передаешь в метод toString() задачу определенного из метода `save()`, а в классах задач теперь есть геттеры, которые возвращают енам типа, по идее, метод `toEnum()` вообще не нужен)
`switch (task.getTypeOfTask())`
Можно еще и сделать к примеру метод-заглушку `getIdEpic()` для классов Task и Epic (return null).
Тогда можно будет свернуть метод в одну строку)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -142,90 +109,51 @@ public static FileBackedTaskManager loadFromFile(File file) {
                 }
-                Task task = fromString(taskString);
-                TypeOfTask typeOfTask = toEnum(task);
+                Task task = Utils.fromString(taskString);
+                TypeOfTask typeOfTask = Utils.toEnum(task);
```
Из `fromString()` возвращается объект определенного типа (Task, Epic или Subtask), поэтому можно получить тип задачи через геттер `task.getTypeOfTask()`

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
```diff
@@ -44,53 +44,34 @@ void loadingAnEmptyFileTest() {
+                    "Ошибка востановления задач");
+            Assertions.assertEquals(fileManager1.getAllEpics(), fileManager2.getAllEpics(),
+                    "Ошибка востановления эпиков");
+            Assertions.assertEquals(fileManager2.getAllSubtasks(), fileManager2.getAllSubtasks(),
+                    "Ошибка востановления подзадач");
```
Обрати внимание, что тут сравниваются списки одинаковых менеджеров)
Поэтому я не рекомендую использовать индексы по типу 1, 2 и тд. в названиях переменных

Mur-Muring committed:
7 Спринт 3/5

Оптимизировать методы в утилитарном классе

Ivanov-s-tmn commented on src/model/Epic.java:
```diff
@@ -23,6 +23,10 @@ public TypeOfTask getTypeOfTask() {
         return TypeOfTask.EPIC;
     }
 
+    public Integer getIdEpic() {
+        return null;
+    }
```
Достаточно заглушки в классе Task, а Epic ее унаследует. В классе Subtask уже будет переопределяться и возвращать epicId
